iroh-one/src/config.rs

```diff
@@ -0,0 +1,162 @@
+use anyhow::Result;
+use axum::http::header::*;
+use config::{ConfigError, Map, Source, Value};
+
+use iroh_metrics::config::Config as MetricsConfig;
+use iroh_p2p::Libp2pConfig;
+use iroh_rpc_client::Config as RpcClientConfig;
+#[cfg(feature = "uds-gateway")]
+use iroh_rpc_types::Addr;
+use iroh_util::insert_into_config_map;
+use serde::{Deserialize, Serialize};
+use std::path::PathBuf;
+
+/// CONFIG_FILE_NAME is the name of the optional config file located in the iroh home directory
+pub const CONFIG_FILE_NAME: &str = "one.config.toml";
+/// ENV_PREFIX should be used along side the config field name to set a config field using
+/// environment variables
+/// For example, `IROH_ONE_PORT=1000` would set the value of the `Config.port` field
+pub const ENV_PREFIX: &str = "IROH_ONE";
+pub const DEFAULT_PORT: u16 = 9050;
+
+/// The configuration includes gateway, store and p2p specific items
+/// as well as the common rpc & metrics ones.
+#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
+pub struct Config {
+    // Gateway specific configuration.
+    pub gateway: iroh_gateway::config::Config,
+    // Store specific configuration.
+    pub store: iroh_store::config::Config,
+    // P2P specific configuration.
+    pub p2p: iroh_p2p::config::Config,
+
+    /// rpc addresses for the gateway & addresses for the rpc client to dial
+    pub rpc_client: RpcClientConfig,
+    /// metrics configuration
+    pub metrics: MetricsConfig,
+    /// Path for the UDS socket for the gateway.
+    #[cfg(feature = "uds-gateway")]
+    pub uds_path: Option<PathBuf>,
+}
+
+impl Config {
+    pub fn new(
+        gateway: iroh_gateway::config::Config,
+        store: iroh_store::config::Config,
+        p2p: iroh_p2p::config::Config,
+        rpc_client: RpcClientConfig,
+        #[cfg(feature = "uds-gateway")] uds_path: Option<PathBuf>,
+    ) -> Self {
+        Self {
+            gateway,
+            store,
+            p2p,
+            rpc_client,
+            metrics: MetricsConfig::default(),
+            #[cfg(feature = "uds-gateway")]
+            uds_path,
+        }
+    }
+
+    // When running in ipfsd mode, the resolver will use memory channels to
+    // communicate with the p2p and store modules.
+    // The gateway itself is exposing a UDS rpc endpoint to be also usable
+    // as a single entry point for other system services if feature enabled.
+    pub fn default_ipfsd() -> RpcClientConfig {
+        #[cfg(feature = "uds-gateway")]
+        let path: PathBuf =
+            format!("{}", std::env::temp_dir().join("ipfsd.gateway").display()).into();
```

Comment:
please use `tempdir`, it is more reliable

Reply:
also, no need for the `format` call,

Reply:
That won't do anything sensible on Android though. Maybe it's best to make that path part of the config.

Reply:
@fabricedesre just a note that this is the default and is overridden by `--gateway-uds-path` if set.